src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography;
+
+namespace System.Net.Http
+{
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+
+        private sealed class ClientCertProvider : IDisposable
+        {
+            internal readonly string _certificateFile;
+            internal readonly string _password;
+
+            private ClientCertProvider(X509Certificate2Collection certs)
+            {
+                Debug.Assert(certs.Count >0);
+                try
+                {
+                    _password = GeneratePassword();
+                    byte[] pfxData = certs.Export(X509ContentType.Pkcs12, _password);
+                    _certificateFile = Path.GetTempFileName();
+                    File.WriteAllBytes(_certificateFile, pfxData);
+                }
+                catch (Exception e)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, e);
+                }
+            }
+
+            public void Dispose()
+            {
+                if (_certificateFile == null)
+                {
+                    return;
+                }
+                try
+                {
+                    File.Delete(_certificateFile);
+                }
+                catch (IOException e)
+                {
+                    VerboseTrace("Deleting temporary file:" + e);
+                }
+            }
+
+            internal static ClientCertProvider Create(ClientCertificateOption option, string uriScheme)
+            {
+                if (option == ClientCertificateOption.Manual || !string.Equals(uriScheme, UriSchemeHttps))
+                {
+                    return null;
+                }
+
+                X509Certificate2Collection clientCerts = new X509Certificate2Collection();
+                using (var myStore = new X509Store())
+                {
+                    myStore.Open(OpenFlags.ReadOnly);
+                    X509Certificate2Collection certificates = myStore.Certificates;
+
+                    certificates = certificates.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, true);
```

Comment:
> There are times when one looks at code and is compelled to ask "did you test this?". In this case, I'm pretty sure the answer has to be "no", because I haven't written Find on Unix yet: It always emits the empty set.

:-)  Truly. But give me some credit please. This was tested with a separate default constructor (with hard coded file names). 

It is not just that the  Find is not implemented.  X509 store on OSX does not even list the cert that are self-signed (the certs that I could test with). 

Adding a cert into X509 store before calling handler too does not work.

I have only two options - include the code that I used to test and explain it and some how ensure that no one merges with the test code in it or wait for some one to point out so that I can explain. I chose the latter. :)

The things I need to test are - whether libcurl accepts a valid P12 file, that the call succeeds when the right cert is used and fails when the wrong cert is used and that the file gets deleted at the end of the call. I assure you they are all tested.
 
> This flow would be incredibly expensive, anyways. It would create two additional collections and the found certs in each collection would be cloned for each call to Find.
This should really just be one for/foreach doing all three tests for performance reasons.

I see. This is essentially a verbatim copy of the code in [winhttpHandler](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L1633) except that the open call has only one flag.

     myStore.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);

results in an error (and I noticed that OpenExistingOnly is expected to thow - there is even a test case for it).
That is the only difference between winhttp handler and this.

I will change as you say - collect certs in one iteration.